Add preupdate hook

aschey opened the pull request:
This is a continuation of the [previous PR](https://github.com/rusqlite/rusqlite/pull/897) by @MidasLamb (thank you!)

Per feedback on the original proposal, this adds a new feature flag (`preupdate_hook`) which will enable the preupdate API. I moved the hooks implementation into its own folder so the preupdate hook can be easily feature-flagged separately. Let me know if you'd like this to be structured differently and I'd be happy to make any adjustments.

gwenn commented on Cargo.toml:
```diff
@@ -95,6 +96,7 @@ modern-full = [
     "extra_check",
     "functions",
     "hooks",
+    "preupdate_hook",
```
Please don't touch `modern-full` because most of the time SQLITE_ENABLE_PREUPDATE_HOOK is not activated by default. So if you use `modern-full` and link against the SQLite3 library installed on your OS, the build will probably fail.

aschey replied:
Good to know, thanks! Fixed.

gwenn commented on src/hooks/preupdate_hook.rs:
```diff
@@ -0,0 +1,333 @@
+    #[inline]
+    pub fn preupdate_hook<'c, F>(&'c self, hook: Option<F>)
+    where
+        F: FnMut(Action, &str, &str, &PreUpdateCase) + Send + 'c,
```
The `'c` (instead of `'static`) seems wrong to me but I am not sure.

aschey replied:
Yes, you're correct. This was a remnant from the original PR where all the other hooks had lifetimes attached. Missed that when I was updating this, thanks!

gwenn commented on src/hooks/preupdate_hook.rs:
```diff
@@ -0,0 +1,333 @@
+
+impl InnerConnection {
+    #[inline]
+    pub fn remove_preupdate_hook(&mut self) {
```
`remove_preupdate_hook` seems useless / redundant to me.

gwenn commented:
I am afraid that I should reject this PR because you haven't fork the previous PR #897.
We should see @MidasLamb contributions in git commits.

aschey commented:
> I am afraid that I should reject this PR because you haven't fork the previous PR #897. We should see @MidasLamb contributions in git commits.

Apologies, thanks for pointing this out. I had merged in the original PR, but I think I accidentally lost the original commit history during a rebase. I wasn't paying attention and missed that. I reapplied the changes on top of the original PR now.

gwenn commented:
Must read / check: https://github.com/rusqlite/rusqlite/issues/1048
Nice to have a compilation failure test like:  https://github.com/rusqlite/rusqlite/pull/1052 (without UnwindSafe / RefUnwindSafe)

aschey commented:
> Must read / check: #1048 Nice to have a compilation failure test like: #1052 (without UnwindSafe / RefUnwindSafe)

Got it, thanks. It needs an explicit `'static` lifetime on the callback. Fixed that and added a complication failure test as mentioned.

aschey commented:
@gwenn I just realized that the `preupdate_hook` features would never get tested in CI since it's not automatically enabled by `bundled-full`. I updated the github actions config to also enable it wherever it currently enables `bundled-full`, `session`, and `buildtime_bindgen`. Let me know if you'd like to handle this differently.

gwenn commented:
Thanks
